From a3cd2c210bae4c3df6fb48d2e06e4e1c8d7ba5fb Mon Sep 17 00:00:00 2001
From: Phantom X <PhantomX@users.noreply.github.com>
Date: Sun, 6 Mar 2022 03:49:26 -0300
Subject: [PATCH] revert logging: Convert `backend_thread` into an
 `std::jthread`

This is preventing Qt gui process to stop
---
 src/common/logging/backend.cpp | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/src/common/logging/backend.cpp b/src/common/logging/backend.cpp
index f1c9ed6..c51c05b 100755
--- a/src/common/logging/backend.cpp
+++ b/src/common/logging/backend.cpp
@@ -218,17 +218,19 @@ private:
     Impl(const std::filesystem::path& file_backend_filename, const Filter& filter_)
         : filter{filter_}, file_backend{file_backend_filename} {}
 
-    ~Impl() = default;
+    ~Impl() {
+        StopBackendThread();
+    }
 
     void StartBackendThread() {
-        backend_thread = std::jthread([this](std::stop_token stop_token) {
+        backend_thread = std::thread([this] {
             Common::SetCurrentThreadName("yuzu:Log");
             Entry entry;
             const auto write_logs = [this, &entry]() {
                 ForEachBackend([&entry](Backend& backend) { backend.Write(entry); });
             };
-            while (!stop_token.stop_requested()) {
-                entry = message_queue.PopWait(stop_token);
+            while (!stop.stop_requested()) {
+                entry = message_queue.PopWait(stop.get_token());
                 if (entry.filename != nullptr) {
                     write_logs();
                 }
@@ -242,6 +244,11 @@ private:
         });
     }
 
+    void StopBackendThread() {
+        stop.request_stop();
+        backend_thread.join();
+    }
+
     Entry CreateEntry(Class log_class, Level log_level, const char* filename, unsigned int line_nr,
                       const char* function, std::string&& message) const {
         using std::chrono::duration_cast;
@@ -276,7 +283,8 @@ private:
     ColorConsoleBackend color_console_backend{};
     FileBackend file_backend;
 
-    std::jthread backend_thread;
+    std::stop_source stop;
+    std::thread backend_thread;
     MPSCQueue<Entry, true> message_queue{};
     std::chrono::steady_clock::time_point time_origin{std::chrono::steady_clock::now()};
 };
-- 
2.35.1

